staticd/static_zebra.c

```diff
@@ -0,0 +1,356 @@
+		return 0;
+
+	static_ifindex_update(ifp, true);
+	return 0;
+}
+
+static int interface_delete(int command, struct zclient *zclient,
+			    zebra_size_t length, vrf_id_t vrf_id)
+{
+	struct interface *ifp;
+	struct stream *s;
+
+	s = zclient->ibuf;
+	/* zebra_interface_state_read () updates interface structure in iflist
+	 */
+	ifp = zebra_interface_state_read(s, vrf_id);
+
+	if (ifp == NULL)
+		return 0;
+
+	if_set_index(ifp, IFINDEX_INTERNAL);
+
+	static_ifindex_update(ifp, false);
+	return 0;
+}
+
+static int interface_address_add(int command, struct zclient *zclient,
+				 zebra_size_t length, vrf_id_t vrf_id)
+{
+	zebra_interface_address_read(command, zclient->ibuf, vrf_id);
+
+	return 0;
+}
+
+static int interface_address_delete(int command, struct zclient *zclient,
+				    zebra_size_t length, vrf_id_t vrf_id)
+{
+	struct connected *c;
+
+	c = zebra_interface_address_read(command, zclient->ibuf, vrf_id);
+
+	if (!c)
+		return 0;
+
+	connected_free(c);
+	return 0;
+}
+
+static int interface_state_up(int command, struct zclient *zclient,
+			      zebra_size_t length, vrf_id_t vrf_id)
+{
+	zebra_interface_if_lookup(zclient->ibuf);
+
+	return 0;
+}
+
+static int interface_state_down(int command, struct zclient *zclient,
+				zebra_size_t length, vrf_id_t vrf_id)
+{
+	zebra_interface_state_read(zclient->ibuf, vrf_id);
+
+	return 0;
+}
+
+static int route_notify_owner(int command, struct zclient *zclient,
+			      zebra_size_t length, vrf_id_t vrf_id)
+{
+	struct prefix p;
+	enum zapi_route_notify_owner note;
+	uint32_t table_id;
+	char buf[PREFIX_STRLEN];
+
+	prefix2str(&p, buf, sizeof(buf));
+
+	if (!zapi_route_notify_decode(zclient->ibuf, &p, &table_id, &note))
+		return -1;
+
+	switch (note) {
+	case ZAPI_ROUTE_FAIL_INSTALL:
+		break;
+	case ZAPI_ROUTE_BETTER_ADMIN_WON:
+		break;
+	case ZAPI_ROUTE_INSTALLED:
+		break;
+	case ZAPI_ROUTE_REMOVED:
+		break;
+	case ZAPI_ROUTE_REMOVE_FAIL:
+		break;
+	}
+
+	return 0;
+}
+static void zebra_connected(struct zclient *zclient)
+{
+	zclient_send_reg_requests(zclient, VRF_DEFAULT);
+}
+
+
+static int static_zebra_nexthop_update(int command, struct zclient *zclient,
+				       zebra_size_t length, vrf_id_t vrf_id)
+{
+	struct zapi_route nhr;
+	afi_t afi = AFI_IP;
+
+	zapi_nexthop_update_decode(zclient->ibuf, &nhr);
+
+	if (nhr.prefix.family == AF_INET6)
+		afi = AFI_IP6;
+
+	static_nht_update(&nhr.prefix, nhr.nexthop_num, afi, vrf_id);
+	return 1;
+}
+
+static void static_zebra_capabilities(struct zclient_capabilities *cap)
+{
+	mpls_enabled = cap->mpls_enabled;
+}
+
+void static_zebra_nht_register(struct static_route *si, bool reg)
+{
+	uint32_t cmd;
+	struct prefix p;
+
+	cmd = (reg) ?
+		ZEBRA_NEXTHOP_REGISTER : ZEBRA_NEXTHOP_UNREGISTER;
+
+	if (si->nh_registered && reg)
+		return;
+
+	if (!si->nh_registered && !reg)
+		return;
+
+	memset(&p, 0, sizeof(p));
+	switch (si->type) {
+	case STATIC_IPV4_GATEWAY_IFNAME:
+	case STATIC_IFNAME:
+	case STATIC_BLACKHOLE:
+	case STATIC_IPV6_GATEWAY_IFNAME:
```

Comment:
I disagree, we were not registering NEXTHOP_TYPE_IPV4_IFNAME and NEXTHOP_TYPE_IPV6_IFNAME before.  They can have no recursion since we are fully specifying the gateway and ifindex.